src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,795 @@
+}
+
+JfrCPUTimeTraceQueue::JfrCPUTimeTraceQueue(u4 capacity) : _capacity(capacity), _head(0), _lost_samples(0) {
+  _data = JfrCHeapObj::new_array<JfrCPUTimeSampleRequest>(capacity);
+}
+
+JfrCPUTimeTraceQueue::~JfrCPUTimeTraceQueue() {
+  JfrCHeapObj::free(_data, _capacity * sizeof(JfrCPUTimeSampleRequest));
+}
+
+bool JfrCPUTimeTraceQueue::enqueue(JfrCPUTimeSampleRequest& request) {
+  assert(JavaThread::current()->jfr_thread_local()->is_cpu_time_jfr_enqueue_locked(), "invariant");
+  u4 elementIndex;
+  do {
+    elementIndex = Atomic::load_acquire(&_head);
+    if (elementIndex >= _capacity) {
+      return false;
+    }
+  } while (Atomic::cmpxchg(&_head, elementIndex, elementIndex + 1) != elementIndex);
+  _data[elementIndex] = request;
+  return true;
+}
+
+JfrCPUTimeSampleRequest& JfrCPUTimeTraceQueue::at(u4 index) {
+  assert(index < _head, "invariant");
+  return _data[index];
+}
+
+volatile u4 _lost_samples_sum = 0;
+
+u4 JfrCPUTimeTraceQueue::size() const {
+  return Atomic::load(&_head);
+}
+
+void JfrCPUTimeTraceQueue::set_size(u4 size) {
+  Atomic::release_store(&_head, size);
+}
+
+u4 JfrCPUTimeTraceQueue::capacity() const {
+  return _capacity;
+}
+
+void JfrCPUTimeTraceQueue::set_capacity(u4 capacity) {
+  _head = 0;
+  JfrCPUTimeSampleRequest* new_data = JfrCHeapObj::new_array<JfrCPUTimeSampleRequest>(capacity);
+  JfrCHeapObj::free(_data, _capacity * sizeof(JfrCPUTimeSampleRequest));
+  _data = new_data;
+  _capacity = capacity;
+}
+
+bool JfrCPUTimeTraceQueue::is_full() const {
+  return Atomic::load_acquire(&_head) >= _capacity;
+}
+
+bool JfrCPUTimeTraceQueue::is_empty() const {
+  return Atomic::load_acquire(&_head) == 0;
+}
+
+s4 JfrCPUTimeTraceQueue::lost_samples() const {
+  return Atomic::load_acquire(&_lost_samples);
+}
+
+void JfrCPUTimeTraceQueue::increment_lost_samples() {
+  Atomic::inc(&_lost_samples_sum);
+  Atomic::inc(&_lost_samples);
+}
+
+u4 JfrCPUTimeTraceQueue::get_and_reset_lost_samples() {
+  s4 lost_samples = Atomic::load_acquire(&_lost_samples);
+  while (Atomic::cmpxchg(&_lost_samples, lost_samples, 0) != lost_samples) {
+    lost_samples = Atomic::load_acquire(&_lost_samples);
+  }
+  return lost_samples;
+}
+
+void JfrCPUTimeTraceQueue::ensure_capacity(u4 capacity) {
+  if (capacity != _capacity) {
+    set_capacity(capacity);
+  }
+}
+
+void JfrCPUTimeTraceQueue::ensure_capacity_for_period(u4 period_millis) {
+  u4 capacity = CPU_TIME_QUEUE_CAPACITY;
+  if (period_millis > 0 && period_millis < 10) {
+    capacity = (u4) ((double) capacity * 10 / period_millis);
+  }
+  ensure_capacity(capacity);
+}
+
+void JfrCPUTimeTraceQueue::clear() {
+  Atomic::release_store(&_head, (u4)0);
+}
+
+static int64_t compute_sampling_period(double rate);
+
+class JfrCPUTimeThreadSampler : public NonJavaThread {
+  friend class JfrCPUTimeThreadSampling;
+ private:
+  Semaphore _sample;
+  NonJavaThread* _sampler_thread;
+  double _rate;
+  bool _autoadapt;
+  volatile int64_t _current_sampling_period_ns = -1;
```

Comment:
Please initialize all variables in the constructor's initializer list instead of inline in the header file.